File: apps/android/app/src/main/java/com/cortexlumora/lunastories/ui/components/ToastCenter.kt
package com.cortexlumora.lunastories.ui.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.cortexlumora.lunastories.ui.theme.Accent
import com.cortexlumora.lunastories.ui.theme.ALPHA_MUTED
import com.cortexlumora.lunastories.ui.theme.GlowCoral
import com.cortexlumora.lunastories.ui.theme.GlowGold
import com.cortexlumora.lunastories.ui.theme.MiloCream
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.UUID

/**
 * App-scoped, lightweight toast notifications. Mirrors iOS ToastCenter:
 * a single in-flight toast, auto-dismissing after a delay, rendered by
 * `ToastOverlay()` placed once at the app root. Two shapes:
 *  - compact error (e.g. generation rejected / quota hit)
 *  - a larger card with a title + progress bar (e.g. "running low" at >=80%)
 */
enum class ToastStyle {
    Error,
    Warning,
    Info,
    Success;

    val icon: ImageVector
        get() = when (this) {
            Error -> Icons.Filled.Warning
            Warning -> Icons.Filled.Error
            Info -> Icons.Filled.Info
            Success -> Icons.Filled.CheckCircle
        }

    val tint: Color
        get() = when (this) {
            Error -> GlowCoral
            Warning -> GlowGold
            Info -> Accent
            Success -> Color(0xFF66C78C)
        }
}

data class Toast(
    val id: String = UUID.randomUUID().toString(),
    val title: String? = null,
    val message: String,
    val style: ToastStyle = ToastStyle.Error,
    /** Optional progress value, 0..1 — drives the "running low" card. */
    val progress: Float? = null,
)

object ToastCenter {
    private val scope = CoroutineScope(Dispatchers.Main)
    private var dismissJob: Job? = null

    private val _current = MutableStateFlow<Toast?>(null)
    val current: StateFlow<Toast?> = _current.asStateFlow()

    /** Show a toast, auto-dismissing after [durationMs]. Replaces any current one. */
    fun show(
        message: String,
        title: String? = null,
        style: ToastStyle = ToastStyle.Error,
        progress: Float? = null,
        durationMs: Long = 3_000,
    ) {
        _current.value = Toast(title = title, message = message, style = style, progress = progress)
        dismissJob?.cancel()
        dismissJob = scope.launch {
            delay(durationMs)
            _current.value = null
        }
    }

    fun dismiss() {
        dismissJob?.cancel()
        _current.value = null
    }
}

/**
 * Renders the current toast pinned to the top. Place once at the app root
 * (e.g. in the root Box). Only the card captures touches, so it never blocks
 * the UI underneath when idle.
 */
@Composable
fun ToastOverlay(modifier: Modifier = Modifier) {
    val current by ToastCenter.current.collectAsState()

    Box(modifier = modifier.fillMaxWidth().statusBarsPadding()) {
        AnimatedVisibility(
            visible = current != null,
            enter = slideInVertically { -it } + fadeIn(),
            exit = slideOutVertically { -it } + fadeOut(),
            modifier = Modifier.align(Alignment.TopCenter),
        ) {
            current?.let { ToastCard(it) { ToastCenter.dismiss() } }
        }
    }
}

@Composable
private fun ToastCard(toast: Toast, onDismiss: () -> Unit) {
    Row(
        modifier = Modifier
            .padding(horizontal = 16.dp, vertical = 8.dp)
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(Color(0xFF2A2138))
            .border(1.dp, toast.style.tint.copy(alpha = 0.45f), RoundedCornerShape(16.dp))
            .padding(horizontal = 16.dp, vertical = 14.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Icon(
            toast.style.icon,
            contentDescription = null,
            tint = toast.style.tint,
            modifier = Modifier.size(20.dp),
        )

        Column(modifier = Modifier.weight(1f)) {
            toast.title?.let { title ->
                Text(
                    title,
                    color = MiloCream,
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold,
                )
                Spacer(Modifier.height(6.dp))
            }
            Text(
                toast.message,
                color = if (toast.title == null) MiloCream else MiloCream.copy(alpha = 0.75f),
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = if (toast.title == null) FontWeight.Medium else FontWeight.Normal,
            )
            toast.progress?.let { p ->
                Spacer(Modifier.height(8.dp))
                ProgressBar(value = p, tint = toast.style.tint)
            }
        }

        Icon(
            Icons.Filled.Close,
            contentDescription = "Dismiss",
            tint = MiloCream.copy(alpha = ALPHA_MUTED),
            modifier = Modifier
                .size(18.dp)
                .clickable(onClick = onDismiss),
        )
    }
}

@Composable
private fun ProgressBar(value: Float, tint: Color) {
    val clamped = value.coerceIn(0f, 1f)
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(6.dp)
            .clip(RoundedCornerShape(3.dp))
            .background(MiloCream.copy(alpha = 0.15f)),
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth(clamped)
                .height(6.dp)
                .clip(RoundedCornerShape(3.dp))
                .background(tint),
        )
    }
}

File: apps/ios/Luna Stories/ToastCenter.swift
//
//  ToastCenter.swift
//  Luna Stories
//
//  App-scoped, lightweight toast notifications. Inject `ToastCenter` into the
//  environment and render `ToastOverlay()` once at the app root; anywhere can
//  then call `toast.show(...)`. Two shapes:
//   - compact error (e.g. generation rejected / quota hit)
//   - a larger card with a title + progress bar (e.g. "running low" at >=80%)
//

import Observation
import SwiftUI

enum ToastStyle: Equatable {
    case error
    case warning
    case info
    case success

    var icon: String {
        switch self {
        case .error: "exclamationmark.triangle.fill"
        case .warning: "exclamationmark.circle.fill"
        case .info: "info.circle.fill"
        case .success: "checkmark.circle.fill"
        }
    }

    var tint: Color {
        switch self {
        case .error: Color(red: 0.95, green: 0.49, blue: 0.34) // warm coral
        case .warning: Color(red: 0.96, green: 0.73, blue: 0.26) // gold
        case .info: .accentColor
        case .success: Color(red: 0.40, green: 0.78, blue: 0.55)
        }
    }
}

struct Toast: Identifiable, Equatable {
    let id = UUID()
    let title: String?
    let message: String
    let style: ToastStyle
    /// Optional progress bar value, 0...1 — drives the "running low" card.
    let progress: Double?
}

@Observable
@MainActor
final class ToastCenter {
    private(set) var current: Toast?
    private var dismissTask: Task<Void, Never>?

    /// Show a toast, auto-dismissing after `duration` seconds. A new toast
    /// replaces any current one (and resets the timer).
    func show(
        _ message: String,
        title: String? = nil,
        style: ToastStyle = .error,
        progress: Double? = nil,
        duration: Double = 3,
    ) {
        current = Toast(title: title, message: message, style: style, progress: progress)
        dismissTask?.cancel()
        dismissTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(duration))
            guard !Task.isCancelled else { return }
            self?.current = nil
        }
    }

    func dismiss() {
        dismissTask?.cancel()
        current = nil
    }
}

/// Renders the current toast pinned to the top of the screen. Place once at
/// the app root (e.g. in an `.overlay(alignment: .top)`). Only the card itself
/// captures touches, so it never blocks the UI underneath when idle.
struct ToastOverlay: View {
    @Environment(ToastCenter.self) private var toast

    var body: some View {
        if let current = toast.current {
            ToastCard(toast: current) { toast.dismiss() }
                .padding(.horizontal, 16)
                .padding(.top, 8)
                .transition(.move(edge: .top).combined(with: .opacity))
                .animation(.spring(response: 0.4, dampingFraction: 0.85), value: toast.current)
        }
    }
}

private struct ToastCard: View {
    let toast: Toast
    let onDismiss: () -> Void

    @State private var dragOffset: CGFloat = 0

    var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Image(systemName: toast.style.icon)
                .font(.system(size: 18, weight: .semibold))
                .foregroundStyle(toast.style.tint)

            VStack(alignment: .leading, spacing: 6) {
                if let title = toast.title {
                    Text(title)
                        .font(.subheadline.weight(.bold))
                        .foregroundStyle(Color.miloCream)
                }
                Text(toast.message)
                    .font(.subheadline.weight(toast.title == nil ? .medium : .regular))
                    .foregroundStyle(Color.miloCream.opacity(toast.title == nil ? 1 : 0.75))
                    .fixedSize(horizontal: false, vertical: true)

                if let progress = toast.progress {
                    ProgressBar(value: progress, tint: toast.style.tint)
                        .frame(height: 6)
                        .padding(.top, 2)
                }
            }

            Spacer(minLength: 8)

            Button(action: onDismiss) {
                Image(systemName: "xmark")
                    .font(.system(size: 13, weight: .bold))
                    .foregroundStyle(Color.miloCream.opacity(0.6))
                    .padding(6)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Dismiss")
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 14)
        .background(
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .fill(.ultraThinMaterial)
                .overlay(
                    RoundedRectangle(cornerRadius: 16, style: .continuous)
                        .strokeBorder(toast.style.tint.opacity(0.45), lineWidth: 1)
                )
        )
        .shadow(color: Color.black.opacity(0.35), radius: 16, x: 0, y: 8)
        .offset(y: dragOffset)
        // Swipe up to dismiss; other directions snap back.
        .gesture(
            DragGesture()
                .onChanged { value in
                    dragOffset = min(0, value.translation.height)
                }
                .onEnded { value in
                    if value.translation.height < -40 {
                        onDismiss()
                    } else {
                        withAnimation(.spring(response: 0.3, dampingFraction: 0.8)) {
                            dragOffset = 0
                        }
                    }
                }
        )
    }
}

private struct ProgressBar: View {
    let value: Double // 0...1
    let tint: Color

    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(Color.miloCream.opacity(0.15))
                Capsule()
                    .fill(tint)
                    .frame(width: geo.size.width * min(max(value, 0), 1))
            }
        }
    }
}
